src/Security/Authentication/Negotiate/src/NegotiateHandler.cs

```diff
@@ -0,0 +1,334 @@
+                if (_negotiateState.Protocol == "NTLM" && !Options.PersistNtlmCredentials)
+                {
+                    // NTLM was already put in the persitence cache on the prior request so we could complete the handshake.
+                    // Take it out if we don't want it to persist.
+                    Debug.Assert(object.ReferenceEquals(persistence?.State, _negotiateState),
+                        "NTLM is a two stage process, it must have already been in the cache for the handshake to succeed.");
+                    persistence.State = null;
+                    Response.RegisterForDispose(_negotiateState);
+                }
+                else if (_negotiateState.Protocol == "Kerberos" && Options.PersistKerberosCredentials)
+                {
+                    persistence ??= EstablishConnectionPersistence(connectionItems);
+                    Debug.Assert(persistence.State == null, "Complete Kerberos results should only be produced from a new context.");
+                    persistence.State = _negotiateState;
+                }
+
+                // Note we run the Authenticated event in HandleAuthenticateAsync so it is per-request rather than per connection.
+            }
+            catch (Exception ex)
+            {
+                var errorContext = new AuthenticationFailedContext(Context, Scheme, Options) { Exception = ex };
+                await Events.AuthenticationFailed(errorContext);
+
+                if (errorContext.Result != null)
+                {
+                    if (errorContext.Result.Handled)
+                    {
+                        return true;
+                    }
+                    else if (errorContext.Result.Skipped)
+                    {
+                        return false;
+                    }
+                    else if (errorContext.Result.Failure != null)
+                    {
+                        throw new Exception("An error was returned from the AuthenticationFailed event.", errorContext.Result.Failure);
+                    }
+                }
+
+                throw;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Checks if the current request is authenticated and returns the user.
+        /// </summary>
+        /// <returns></returns>
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            if (!_requestProcessed)
+            {
+                throw new InvalidOperationException("AuthenticateAsync must not be called before the UseAuthentication middleware.");
+            }
+
+            if (IsHttp2)
+            {
+                // Not supported. We don't throw because Negotiate may be set as the default auth
+                // handler on a server that's running HTTP/1 and HTTP/2. We'll challenge HTTP/2 requests
+                // that require auth and they'll downgrade to HTTP/1.1.
+                return AuthenticateResult.NoResult();
```

Comment:
Is this behavior of downgrading to 1.1 when challenged by spec or convention? It sounds like it's the client which is responsible for the downgrade to 1.1? If so, then what happens if a client decides it knows how to do Negotiate auth over 2.0?

Reply:
I haven't found a spec for it, but all of the clients I tested did do the downgrade except SocketsHttpHandler (it just returned the 401 to the caller).

If a client tries to do Negotiate over HTTP/2 we won't honor it and they'll get a second 401 challenge when they reach the restricted resource.